src/shared/Microsoft.Git.CredentialManager/Authentication/OAuth/OAuth2Client.cs

```diff
@@ -0,0 +1,440 @@
+using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.Git.CredentialManager.Authentication.OAuth.Json;
+using Newtonsoft.Json;
+
+namespace Microsoft.Git.CredentialManager.Authentication.OAuth
+{
+    /// <summary>
+    /// Represents an OAuth2 client application that can perform the basic flows outlined in RFC 6749,
+    /// as well as extensions such as OAuth2 Device Authorization Grant (RFC 8628).
+    /// </summary>
+    public interface IOAuth2Client
+    {
+        /// <summary>
+        /// Retrieve an authorization code grant using a user agent.
+        /// </summary>
+        /// <param name="scopes">Scopes to request.</param>
+        /// <param name="browser">User agent to use to start the authorization code grant flow.</param>
+        /// <returns>Authorization code.</returns>
+        Task<string> GetAuthorizationCodeAsync(IEnumerable<string> scopes, IWebBrowser browser);
+
+        /// <summary>
+        /// Retrieve a device code grant.
+        /// </summary>
+        /// <param name="scopes">Scopes to request.</param>
+        /// <exception cref="InvalidOperationException">Thrown if the client has not been configured with a device authorization endpoint.</exception>
+        /// <returns>Device code grant result.</returns>
+        Task<OAuth2DeviceCodeResult> GetDeviceCodeAsync(IEnumerable<string> scopes);
+
+        /// <summary>
+        /// Exchange an authorization code acquired from <see cref="GetAuthorizationCodeAsync"/> for an access token.
+        /// </summary>
+        /// <param name="authorizationCode">Authorization code.</param>
+        /// <returns>Token result.</returns>
+        Task<OAuth2TokenResult> GetTokenByAuthorizationCodeAsync(string authorizationCode);
+
+        /// <summary>
+        /// Use a refresh token to get a new access token.
+        /// </summary>
+        /// <param name="refreshToken">Refresh token.</param>
+        /// <returns>Token result.</returns>
+        Task<OAuth2TokenResult> GetTokenByRefreshTokenAsync(string refreshToken);
+
+        /// <summary>
+        /// Exchange a device code grant acquired from <see cref="GetDeviceCodeAsync"/> for an access token.
+        /// </summary>
+        /// <param name="deviceCodeResult">Device code grant result.</param>
+        /// <returns>Token result.</returns>
+        Task<OAuth2TokenResult> GetTokenByDeviceCodeAsync(OAuth2DeviceCodeResult deviceCodeResult);
+    }
+
+    public class OAuth2Client : IOAuth2Client
+    {
+        private readonly HttpClient _httpClient;
+        private readonly OAuth2ServerEndpoints _endpoints;
+        private readonly Uri _redirectUri;
+        private readonly string _clientId;
+        private readonly string _clientSecret;
+
+        public OAuth2Client(HttpClient httpClient, OAuth2ServerEndpoints endpoints, string clientId, Uri redirectUri = null, string clientSecret = null)

```

Comment:
Should we document which parameters with defaults are required for a particular flow? For example, I can construct an `OAuth2Client` with no `redirectUri`, but then I cannot call `GetAuthorizationCodeAsync` on that client.

Reply:
This is an error in my logic! Redirect URI does not have to be specified for this flow to work.. the server should use the default configured redirect URI.